python-spec/src/somacore/coordinates.py

```diff
@@ -0,0 +1,491 @@
+"""Definitions of types related to coordinate systems."""
+
+import abc
+import collections.abc
+from typing import Any, Optional, Sequence, Tuple, Union
+
+import numpy as np
+import numpy.typing as npt
+
+
+class Axis:
+    """A description of an axis of a coordinate system
+
+    Args:
+        name: Name of the axis.
+        unit_name: Optional string name for the units. Defaults to ``None``.
+        unit_scale: Optional scale for units. Defaults to ``None``.
+
+    Lifecycle: experimental
+    """
+
+    __slots__ = ("_name", "_unit_name", "_unit_scale")
+
+    def __init__(
+        self,
+        name: str,
+        unit_name: Optional[str] = None,
+        unit_scale: Optional[np.float64] = None,
+    ):
+        self._name: str = name
+        self._unit_name: Optional[str] = unit_name
+        self._unit_scale: Optional[np.float64] = unit_scale
+
+    @property
+    def name(self) -> str:
+        """Name of the axis."""
+        return self._name
+
+    @property
+    def unit_name(self) -> Optional[str]:
+        """Optional string name for the physical quantity the units are measured in."""
+        return self._unit_name
+
+    @property
+    def unit_scale(self) -> Optional[np.float64]:
+        """Optional scale for the physical quantity the units."""
+        return self._unit_scale
+
+
+class CoordinateSpace(collections.abc.Sequence):
```

Comment:
This can also be an `attrs` class, with a `validator`:

```
@attrs.define(frozen=True)
class CoordinateSpace(collections.abc.Sequence[Axis]):
  """..."""

  axes: Tuple[Axis, ...] = attrs.field(converter=tuple)

  @axes.validator
  def _validate(self, _, axes: Tuple[Axis, ...]) -> None:
    if not axes:
      raise ValueError("at least one")
    if len(set(...)) != len(axes):
      raise ValueError("unique")
```

This will also give you `__repr__` for free.

Reply:
Sweet! That validator is nice -  I can see already why you like `attrs` over `dataclass`.